File: project_euler/02/027.go
// Copyright [2017] <dmnsn7@gmail.com>

package main

import "fmt"

const cA = 1000
const cB = 1000
const cN = 2001000

func primeSieve(n int) (map[int]int, []int) {
	isPrime := make([]bool, n)
	for i := 0; i < n; i++ {
		isPrime[i] = true
	}
	mPrime, vPrime := map[int]int{}, []int{}
	for i := 2; i < n; i++ {
		if isPrime[i] {
			mPrime[i] = 0
			vPrime = append(vPrime, i)
		}
		for j := 0; j < len(vPrime) && vPrime[j]*i < n; j++ {
			isPrime[vPrime[j]*i] = false
			if i%vPrime[j] == 0 {
				break
			}
		}
	}

	return mPrime, vPrime
}

func main() {
	mPrime, _ := primeSieve(cN)

	maxLeng, multi := 0, 0
	for i := -cA + 1; i <= cA-1; i++ {
		for j := -cB + 1; j <= cB-1; j++ {
			leng := -1
			for k := 0; leng == -1; k++ {
				num := k*k + i*k + j
				if _, exist := mPrime[num]; !exist {
					leng = k
				}
			}

			if leng > maxLeng {
				maxLeng = leng
				multi = i * j
			}
		}
	}

	fmt.Printf("%d\n", multi)
}

File: project_euler/02/027.py
#!/usr/bin/env python3
'''Copyright [2017] <dmnsn7@gmail.com>'''

A = 1000
B = 1000
N = 2001000


def prime_sieve(num_n):
    is_prime = [True for _ in range(num_n)]
    prime = set()
    for i in range(2, num_n):
        if is_prime[i]:
            prime.add(i)

        for pri in prime:
            if pri * i >= num_n:
                break
            is_prime[pri * i] = False
            if i % pri == 0:
                break

    return prime


def main():
    prime = prime_sieve(N)
    max_leng, multi = 0, 0
    for i in range(-A + 1, A):
        for j in range(-B + 1, B):
            leng, k = -1, 0
            while leng == -1:
                num = k * k + i * k + j
                if num not in prime:
                    leng = k
                k += 1
            if leng > max_leng:
                max_leng = leng
                multi = i * j

    print(multi)


if __name__ == '__main__':
    main()
